pkg/util/limiter/utilization.go

```diff
@@ -0,0 +1,330 @@
+	cpuR, err := readFileNoStat(cpuPathV1)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer cpuR.Close()
+	// CPU time in nanoseconds
+	var cpuTime uint64
+	if _, err := fmt.Fscanf(cpuR, "%d", &cpuTime); err != nil {
+		return 0, 0, errors.Wrapf(err, "failed scanning %s", cpuPathV1)
+	}
+
+	return float64(cpuTime) / 1e09, memUtil, nil
+}
+
+func newCgroupV1Scanner() (utilizationScanner, bool) {
+	if _, err := os.Stat(memPathV1); err != nil {
+		return cgroupV1Scanner{}, false
+	}
+	if _, err := os.Stat(cpuPathV1); err != nil {
+		return cgroupV1Scanner{}, false
+	}
+	return cgroupV1Scanner{}, true
+}
+
+type procfsScanner struct {
+	proc procfs.Proc
+}
+
+func (s procfsScanner) String() string {
+	return "/proc"
+}
+
+func (s procfsScanner) Scan() (float64, uint64, error) {
+	ps, err := s.proc.Stat()
+	if err != nil {
+		return 0, 0, errors.Wrap(err, "failed to get process stats")
+	}
+
+	return ps.CPUTime(), uint64(ps.ResidentMemory()), nil
+}
+
+func newProcfsScanner() (utilizationScanner, bool) {
+	p, err := procfs.Self()
+	if err != nil {
+		return procfsScanner{}, false
+	}
+
+	return procfsScanner{
+		proc: p,
+	}, true
+}
+
+// UtilizationBasedLimiter is a Service offering limiting based on CPU and memory utilization.
+//
+// The respective CPU and memory utilization limits are configurable.
+type UtilizationBasedLimiter struct {
+	services.Service
+
+	logger             log.Logger
+	utilizationScanner utilizationScanner
+
+	// Memory limit in bytes
+	memoryLimit uint64
+	// CPU limit in cores
+	cpuLimit float64
+	// Last CPU time counter
+	lastCPUTime float64
+	// The time of the last update
+	lastUpdate     time.Time
+	movingAvg      *math.EwmaRate
+	LimitingReason atomic.String
+}
+
+// NewUtilizationBasedLimiter returns a UtilizationBasedLimiter configured with cpuLimit and memoryLimit.
+func NewUtilizationBasedLimiter(cpuLimit float64, memoryLimit uint64, logger log.Logger) *UtilizationBasedLimiter {
+	// Calculate alpha for a minute long window
+	// https://github.com/VividCortex/ewma#choosing-alpha
+	alpha := 2 / (60/resourceUtilizationUpdateInterval.Seconds() + 1)
+	l := &UtilizationBasedLimiter{
+		logger:      logger,
+		cpuLimit:    cpuLimit,
+		memoryLimit: memoryLimit,
+		// Use a minute long window, each sample being a second apart
+		movingAvg: math.NewEWMARate(alpha, resourceUtilizationUpdateInterval),
+	}
+	l.Service = services.NewTimerService(resourceUtilizationUpdateInterval, l.init, l.update, nil)
+	return l
+}
+
+func (l *UtilizationBasedLimiter) init(_ context.Context) error {
+	s, ok := newCgroupV2Scanner()
+	if ok {
+		l.utilizationScanner = s
+		return nil
+	}
+	s, ok = newCgroupV1Scanner()
+	if ok {
+		l.utilizationScanner = s
+		return nil
+	}
+	s, ok = newProcfsScanner()
+	if ok {
+		l.utilizationScanner = s
+		return nil
+	}
+
+	return fmt.Errorf("unable to detect CPU/memory utilization, unsupported platform")
+}
+
+func (l *UtilizationBasedLimiter) update(ctx context.Context) error {
```

Comment:
I would like to see a unit test for this. The easiest way to test it is:

- Call `NewUtilizationBasedLimiter()` but do **not** start the service
- Override `utilizationScanner` with a mock implementation for which you control the values returned by `Scan()`
- Call `update()` as many times as you want, to see how `LimitingReason` and `movingAvg` changes (I think it's particularly interesting making sure the CPU utilization is computed correctly)

BTW, you will need to mock time. You can just move all this logic to another function which takes in input `time.Time` (that's the function you test) and then `update()` will just call the new function passing `time.Now()`.

Reply:
Thanks, agree there should be a test. I just didn't want to write one before everything is agreed on :)